File: lectrick/tiles/power_source.py
from .base import Tile


class PowerSource(Tile):
    TILE_TYPE = "POWER SOURCE"
    SAMPLE_CHARS = ["🗲", "⚡", "⌁"]
    POWER_LEVEL = 255

    def __init__(self, x, y, tiles):
        super().__init__(x, y, tiles)
        self.has_sent_power = False

    def start_turn(self):
        if not self.has_sent_power:
            for dx in range(-1, 2):
                for dy in range(-1, 2):
                    if dx == 0 and dy == 0:
                        continue
                    target_x = self.x + dx
                    target_y = self.y + dy
                    try:
                        tile_index = self.tiles.index((target_x, target_y))
                    except ValueError:
                        continue
                    self.tiles[tile_index].accept_energy(self.POWER_LEVEL, self.x, self.y)
            self.has_sent_power = True

File: web/src/tiles/powerSource.ts
import {Tile} from './base';

class PowerSourceTile extends Tile {
    TILE_TYPE = 'POWER SOURCE';
    SAMPLE_CHARS = ["🗲", "⚡", "⌁"];
    POWER = 255;
    hasSentPower: boolean;

    constructor(ctx: ExecutionContext, x: number, y: number) {
        super(ctx, x, y);
        this.hasSentPower = false;
    }

    startTurn() {
        if (!this.hasSentPower) {
            for (let dx = 0; dx < 2; dx++) {
                for (let dy = 0; dy < 2; dy++) {
                    if ((dx == 0 && dy == 0) || (dx != 0 && dy != 0)) {
                        continue;
                    }
                    const targetX = this.x + dx;
                    const targetY = this.y + dy;
                    const targetTile = this.ctx.getTile(targetX, targetY);
                    if (targetTile) {
                        targetTile.acceptEnergy(this.POWER, this.x, this.y);
                    }
                }
            }
            this.hasSentPower = true;
        }
    }
}